File: web/src/lib/freshness.test.ts
import { describe, expect, it } from "vitest";
import { classify, formatAge, LIVE, PROCESS_MAX_AGE_MS, STALE, UNKNOWN } from "./freshness";

const nowMs = 1_700_000_010_000;

describe("classify", () => {
  it("LIVE when ALIVE and recent", () => {
    expect(classify(nowMs, 1_700_000_005_000, "ALIVE")).toBe(LIVE);
  });

  it("LIVE when ALIVE and age == 10s", () => {
    expect(classify(nowMs, 1_700_000_000_000, "ALIVE")).toBe(LIVE);
  });

  it("STALE when ALIVE and old", () => {
    expect(classify(nowMs, 1_699_999_999_000, "ALIVE")).toBe(STALE);
  });

  it("STALE when FAILED even if recent", () => {
    expect(classify(nowMs, 1_700_000_009_000, "FAILED")).toBe(STALE);
  });

  it("UNKNOWN when no timestamp", () => {
    expect(classify(nowMs, 0, "ALIVE")).toBe(UNKNOWN);
  });

  it("UNKNOWN when REVOKED", () => {
    expect(classify(nowMs, 1_700_000_009_000, "REVOKED")).toBe(UNKNOWN);
  });

  it("STALE when SUSPECT even if recent", () => {
    expect(classify(nowMs, 1_700_000_009_000, "SUSPECT")).toBe(STALE);
  });

  it("STALE when LEFT even if recent", () => {
    expect(classify(nowMs, 1_700_000_009_000, "LEFT")).toBe(STALE);
  });

  it("STALE when JOINING with timestamp", () => {
    expect(classify(nowMs, 1_700_000_009_000, "JOINING")).toBe(STALE);
  });

  it("LIVE when ALIVE and age == 40s with process max age", () => {
    expect(classify(nowMs, nowMs - 40_000, "ALIVE", PROCESS_MAX_AGE_MS)).toBe(LIVE);
  });

  it("STALE when ALIVE and age > 40s with process max age", () => {
    expect(classify(nowMs, nowMs - 40_001, "ALIVE", PROCESS_MAX_AGE_MS)).toBe(STALE);
  });

  it("STALE when ALIVE and age == 11s with default node max age", () => {
    expect(classify(nowMs, nowMs - 11_000, "ALIVE")).toBe(STALE);
  });
});

describe("formatAge", () => {
  it("formats 10s age", () => {
    const got = formatAge(1_700_000_010_000, 1_700_000_000_000);
    expect(got.includes("10s ago") || got.includes("10s")).toBe(true);
  });
});

File: internal/freshness/freshness_test.go
package freshness

import (
	"testing"
	"time"
)

var classifyNow = time.UnixMilli(1_700_000_010_000)

func TestClassify_LiveAliveRecent(t *testing.T) {
	if got := Classify(classifyNow, 1_700_000_005_000, "ALIVE"); got != LIVE {
		t.Fatalf("recent ALIVE: got %s want LIVE", got)
	}
	if got := Classify(classifyNow, 1_700_000_000_000, "ALIVE"); got != LIVE {
		t.Fatalf("ALIVE age==10s: got %s want LIVE", got)
	}
}

func TestClassify_StaleAliveOld(t *testing.T) {
	if got := Classify(classifyNow, 1_699_999_999_000, "ALIVE"); got != STALE {
		t.Fatalf("old ALIVE: got %s want STALE", got)
	}
}

func TestClassify_StaleFailedEvenIfRecent(t *testing.T) {
	if got := Classify(classifyNow, 1_700_000_009_000, "FAILED"); got != STALE {
		t.Fatalf("recent FAILED: got %s want STALE", got)
	}
}

func TestClassify_UnknownNoTimestamp(t *testing.T) {
	if got := Classify(classifyNow, 0, "ALIVE"); got != UNKNOWN {
		t.Fatalf("last=0: got %s want UNKNOWN", got)
	}
}

func TestClassify_UnknownRevoked(t *testing.T) {
	if got := Classify(classifyNow, 1_700_000_009_000, "REVOKED"); got != UNKNOWN {
		t.Fatalf("REVOKED: got %s want UNKNOWN", got)
	}
}

func TestClassify_StaleSuspect(t *testing.T) {
	if got := Classify(classifyNow, 1_700_000_009_000, "SUSPECT"); got != STALE {
		t.Fatalf("recent SUSPECT: got %s want STALE", got)
	}
}

func TestClassify_StaleLeft(t *testing.T) {
	if got := Classify(classifyNow, 1_700_000_009_000, "LEFT"); got != STALE {
		t.Fatalf("recent LEFT: got %s want STALE", got)
	}
}

func TestClassify_StaleJoiningWithTimestamp(t *testing.T) {
	if got := Classify(classifyNow, 1_700_000_009_000, "JOINING"); got != STALE {
		t.Fatalf("recent JOINING: got %s want STALE", got)
	}
}
